Remotion/Web/ClientScript/Scripts/SmartPage.ts

```diff
@@ -258,19 +321,42 @@ class SmartPage_Context
     }
   };
 
-  public SmartPage_PageRequestManager_endRequest = function (sender: unknown, args: Sys.WebForms.EndRequestEventArgs)
+  public OverrideErrorMessageForStatusCode (statusCode: number, message: string) : void {
+    let current = this._errorMessageMap.get(statusCode);
+    if (current == undefined) return;
+
+    current.errorMessage = message;
+  }
+
+  public SmartPage_PageRequestManager_endRequest = (sender: unknown, args: Sys.WebForms.EndRequestEventArgs) =>
   {
-    if (args.get_error() != undefined && (args.get_error() as any).httpStatusCode == '500')
+    if (args.get_error() != undefined)
     {
-      const errorMessage = args.get_error().message;
-      args.set_errorHandled(true);
+      let statusCode = (args.get_error() as any).httpStatusCode;
 
-      const errorBody = '<div class="SmartPageErrorBody"><div>' + errorMessage + '</div></div>';
+      let errorDeclaration = this._errorMessageMap.get(statusCode);
 
-      SmartPage_Context.Instance!.ShowMessage("SmartPageServerErrorMessage", errorBody);
+      if (errorDeclaration == undefined || statusCode == 200){
+        this.HandleEndRequestError("undefined", args.get_error().name, "An error has occured.", args, true);
+      }
+      else {
+        this.HandleEndRequestError(statusCode, errorDeclaration.shortDescription, errorDeclaration.errorMessage, args, false);
+      }
     }
   }
 
+  private HandleEndRequestError(statusCode: string, errorDescription: string, errorMessage: string, args: Sys.WebForms.EndRequestEventArgs, showBody: boolean): void
+  {
+    args.set_errorHandled(true);
+
+    let errorBody = '<div class="SmartPageErrorBody"><div>Error ' + statusCode + ', ' + errorDescription + ': ' + errorMessage + '</div>';
+    if (showBody)
+      errorBody += '<iframe sandbox srcdoc="<div>' + args.get_response().get_responseData() + '</div>"></iframe>';
+    errorBody += '</div>';
```

Comment:
We need to discuss what parts of the response we consider "dangerous" and should therefore put in the `iframe`. Also spot the XSS problem :D

Reply:
I presume putting the response data in the iframe is the XSS problem, but shouldn't that be handled by sandbox? Or do we have a problem by simply calling the 'toString' of the response data? And yea, we do need to talk about whats 'dangerous'

Reply:
What is the potential attach vector? And what input could produce it?